bpf: lb: introduce an optimized CT lookup

julianwiedmann opened the pull request:
This splits off the CT improvements from https://github.com/cilium/cilium/pull/22978 for easier reviewing (ie no need to follow all the DSR changes in that PR). They are needed to slim down the instruction count in the nodeport path. 

The main idea is to extract a fully populated CT tuple in the service lookup path *once*, and thus avoid multiple L4 port extractions in the `ct_lookup()` calls. As we need to extract the `dest` port for the Service lookup anyway, we can also extract the `source` port at this time without extra cost.

I expect that we can use the new `ct_lb_lookup()` helpers in more locations over time, and presumably also rename it at that point.

julianwiedmann committed:
bpf: lb: have lb_local() update CT tuple with backend port

The SVC-to-backend DNAT code already updates the .daddr in the CT tuple.
Also update the backend port, so that the nodeport code can use this tuple
as-is to create an CT entry for RevDNAT of non-DSR connections.

Slightly pull down the .daddr update in lb6_local(), to keep the code in
sync with lb4_local() and have the whole tuple update in one place.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: conntrack: introduce an optimized CT lookup for LB

Allow the LB paths to perform a CT lookup without validating the
L3 protocol or extracting the L4 ports. This helps to reduce instruction
count in the nodeport code.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: conntrack: use enum ct_dir for CT_* direction

Enforce proper types.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: lb: de-dup and simplify l4_xlate

Trust that earlier stages have already checked the L4 protocol, and we're
not suddenly processing a request with non-SVC protocol.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: lb: pull ipv4_handle_fragmentation() into lb4_extract_tuple()

We already have a validated IPv4 header, so avoid jumping through
ipv4_ct_extract_l4_ports() for loading the ports.

Stick close to how lb4_extract_l4_port() lays out the code, as we want to
start using lb4_extract_tuple() in more places and the 4.19 verifier is
very picky about it.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: lb: use lb6_extract_tuple() in SVC lookup paths

As part of the nodeport SVC lookup, lb6_extract_key() currently extracts
the packet's destination port. Refactor this code path to use
lb6_extract_tuple() instead, thus loading _both_ L4 ports. We can use this
in a subsequent patch to slim down the CT lookup code.

lb6_extract_key() turns into a much simpler helper that only fills the
lb6_key.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: lb: fine-tune extract_l4_port()

This helper is now only used for IPv4 packets. Remove the IPv6-specific
parts.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: lb: use lb4_extract_tuple() in SVC lookup paths

Same as for lb6_extract_tuple().

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

gentoo-root commented on bpf/lib/ipv4.h:
```diff
@@ -143,7 +143,7 @@ ipv4_handle_fragmentation(struct __ctx_buff *ctx,
 	ret = ctx_load_bytes(ctx, l4_off, ports, 4);
 	if (ret < 0)
-		return ret;
+		return DROP_CT_INVALID_HDR;
```
While this change makes sense to me, I seem to miss the point how it's related to the rest of this commit.

julianwiedmann replied:
Hmm maybe just leftovers from trying to return proper DROP reasons from `lb4_extract_tuple()`. But probably doesn't make sense to keep it now, until we can also do so in its non-`ENABLE_IPV4_FRAGMENTS` variant.

Happy to drop this into a follow-on PR, we have no lack of missing errno translations ;).

julianwiedmann replied:
> Happy to drop this into a follow-on PR, we have no lack of missing errno translations ;).

done

gentoo-root commented on bpf/lib/lb.h:
```diff
@@ -1202,8 +1204,18 @@ lb4_extract_tuple(struct __ctx_buff *ctx, struct iphdr *ip4, int *l4_off,
+
+		if (IS_ERR(ret))
+			return ret;
+		return 0;
```
Why not remove revalidate_data from ipv4_ct_extract_l4_ports, instead of copying its body?

The comment above that revalidate_data mentions it's important for ct_lookup4, so maybe it's worth calling revalidate_data from ct_lookup4 only?

julianwiedmann replied:
I didn't want to walk down the refactor rabbit hole too much - having something that passes the 4.19 verifier in `nodeport_lb4()` was difficult enough :sweat_smile:.

But yes, some follow-on cleanups for the remaining port extraction helpers would be good. And getting rid of such deeply embedded `revalidate_data()` calls would be nice for other reasons too.

gentoo-root commented on bpf/lib/lb.h:
```diff
@@ -1202,8 +1204,18 @@ lb4_extract_tuple(struct __ctx_buff *ctx, struct iphdr *ip4, int *l4_off,
+						(struct ipv4_frag_l4ports *)&tuple->dport,
+						NULL);
+#else
+		ret = l4_load_ports(ctx, *l4_off, &tuple->dport);
```
Given that it's copied from ipv4_ct_extract_l4_ports, but you replaced a raw ctx_load_bytes with l4_load_ports, maybe it's worth to do the same replacement in ipv4_ct_extract_l4_ports as well?

julianwiedmann replied:
ack, `bpf: clean up open-coded users of l4_load_ports()` is already sitting in my cleanups branch. Will upstream once the dust has settled :)

pchaigno reviewed (approved):
LGTM. Last comment is mixing a small logic change with a fair amount of refactoring; probably would have been easier to review if the two were separate. That said, the rest of the commit history is :ok_hand:

ldelossa commented on bpf/bpf_lxc.c:
```diff
@@ -620,26 +620,19 @@ static __always_inline int __tail_handle_ipv6(struct __ctx_buff *ctx)
 		}
 
+		lb6_fill_key(&key, &tuple);
+		csum_l4_offset_and_flags(tuple.nexthdr, &csum_off);
```
Why pull csum_l4_offset_and_flags() out of lb6_fill_key, couldn't it still be called internally?

julianwiedmann replied:
I didn't want to call it `lb6_fill_mix_of_random_stuff()` though ;).

Kidding aside - setting up the `csum_off` should eventually migrate down into `lb_local()` where it's actually needed. Stay tuned ...

ldelossa replied:
👌

ldelossa commented on bpf/lib/lb.h:
```diff
@@ -507,36 +507,28 @@ static __always_inline int lb6_rev_nat(struct __ctx_buff *ctx, int l4_off,
 
-/** Extract IPv6 LB key from packet
+static __always_inline void
+lb6_fill_key(struct lb6_key *key, struct ipv6_ct_tuple *tuple)
```
This is probably out of scope for the PR, but it feels a bit weird that we need to transit thru a CT tuple to extract the key. Couldn't the key simply come from a IPv6 header?

julianwiedmann replied:
> Couldn't the key simply come from a IPv6 header?

That's the approach we want to get *away* from :sweat_smile:. Extract the tuple once, use it as source of truth, adjust if needed.

If we end up with other users that don't have a tuple but still want to fill a key, we can always adjust `lb6_fill_key()` and just pass the needed values manually.

ldelossa replied:
👍

ldelossa reviewed (approved):
LGTM, couple questions.

julianwiedmann committed:
bpf: lb: use new ct_lb_lookup() helpers

As the LB paths are doing their own L3 protocol checks and also extract
the L4 ports on their own, we can switch lb_local() and all the nodeport
code to use the optimized CT lookup helpers.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>
